File: syn-refinements/SumEq.hs
{-@ LIQUID "--higherorder"                         @-}
{-@ LIQUID "--automatic-instances=liquidinstances" @-}

module SumEq where

-- | Recursive implementation -----------------------------------------

{-@ reflect sumTo @-}
{-@ sumTo :: Nat -> Nat @-}
sumTo :: Int -> Int
sumTo n
  | n == 0    = 0
  | otherwise = n + sumTo (n - 1)

-- | Tail-Recursive implementation ------------------------------------

{-@ reflect loop              @-}
{-@ loop :: Nat -> Nat -> Nat @-}
loop :: Int -> Int -> Int
loop n acc
  | n == 0    = acc
  | otherwise = loop (n - 1) (acc + n)

{-@ reflect sumToTR       @-}
{-@ sumToTR :: Nat -> Nat @-}
sumToTR n = loop n 0

-- | Invariant relating the two implementations --------------------

{-@ inv :: n:Nat -> acc:Nat -> { loop n acc == acc + sumTo n } @-}
inv :: Int -> Int -> ()
inv n acc
  | n == 0    = ()
  | otherwise = inv (n - 1) (acc + n)

-- | Equivalence of two implementations ----------------------------

{-@ equiv :: n:Nat -> {sumTo n = sumToTR n} @-}
equiv :: Int -> ()
equiv n = inv n 0

File: papers/syn-refinements/SumEq.ml
(* -- Recursive Summation ------------- *)

let rec sumTo n = if n = 0
                    then 0
                    else n + sumTo (n - 1)


(* -- Tail-Recursive Summation ------------------ *)

let sumToTR n =
  let rec loop n acc = if n == 0
                         then acc
                         else loop (n - 1) (acc + n)
  in
    loop n 0

(* -- Invariant relating the two implementations ----------------------- *)

(*-@  val inv :: n:Nat -> acc:Nat -> { loop n acc == acc + sumTo n }   @-*)

let rec inv n acc = if n == 0
                      then ()
                      else inv (n - 1) (acc + n)

(* -- Equivalence of two implementations ------------------------------- *)

(*-@ val equiv :: n:Nat -> {sumTo n = sumToTR n}                       @-*)

let equiv n = inv n 0
